File: go/parser.go
package ason

import (
	"fmt"
	"strconv"
)

// SchemaField represents a field in the schema.
type SchemaField struct {
	Name     string
	IsArray  bool
	Children []*SchemaField // for nested objects
}

// Parser parses ASON input.
type Parser struct {
	lexer *Lexer
	cur   Token
}

// NewParser creates a new parser for the given input.
func NewParser(input string) *Parser {
	p := &Parser{
		lexer: NewLexer(input),
	}
	p.advance()
	return p
}

func (p *Parser) advance() {
	p.cur = p.lexer.NextToken()
}

func (p *Parser) expect(typ TokenType) error {
	if p.cur.Type != typ {
		return fmt.Errorf("expected %v, got %v at line %d, column %d",
			typ, p.cur.Type, p.cur.Line, p.cur.Column)
	}
	p.advance()
	return nil
}

// Parse parses the ASON input and returns the result.
func (p *Parser) Parse() (*Value, error) {
	// Check if this is schema:data format or just data
	if p.cur.Type == TokenLBrace {
		return p.parseSchemaAndData()
	}
	// Pure data format
	return p.parseValue()
}

func (p *Parser) parseSchemaAndData() (*Value, error) {
	// Parse schema
	schema, err := p.parseSchema()
	if err != nil {
		return nil, err
	}

	// Expect colon
	if err := p.expect(TokenColon); err != nil {
		return nil, err
	}

	// Parse data tuples
	var results []*Value
	for {
		val, err := p.parseDataWithSchema(schema)
		if err != nil {
			return nil, err
		}
		results = append(results, val)

		if p.cur.Type != TokenComma {
			break
		}
		p.advance() // consume comma

		// Check if next is another tuple or end
		if p.cur.Type != TokenLParen {
			break
		}
	}

	if len(results) == 1 {
		return results[0], nil
	}
	return Array(results...), nil
}

func (p *Parser) parseSchema() ([]*SchemaField, error) {
	if err := p.expect(TokenLBrace); err != nil {
		return nil, err
	}

	var fields []*SchemaField
	for p.cur.Type != TokenRBrace && p.cur.Type != TokenEOF {
		field, err := p.parseSchemaField()
		if err != nil {
			return nil, err
		}
		fields = append(fields, field)

		if p.cur.Type == TokenComma {
			p.advance()
		}
	}

	if err := p.expect(TokenRBrace); err != nil {
		return nil, err
	}

	return fields, nil
}

func (p *Parser) parseSchemaField() (*SchemaField, error) {
	if p.cur.Type != TokenIdent {
		return nil, fmt.Errorf("expected identifier at line %d, column %d",
			p.cur.Line, p.cur.Column)
	}

	field := &SchemaField{Name: p.cur.Value}
	p.advance()

	// Check for array marker []
	if p.cur.Type == TokenLBracket {
		p.advance()
		field.IsArray = true
		// Check for nested object schema in array
		if p.cur.Type == TokenLBrace {
			children, err := p.parseSchema()
			if err != nil {
				return nil, err
			}
			field.Children = children
		}
		if err := p.expect(TokenRBracket); err != nil {
			return nil, err
		}
	} else if p.cur.Type == TokenLBrace {
		// Nested object
		children, err := p.parseSchema()
		if err != nil {
			return nil, err
		}
		field.Children = children
	}

	return field, nil
}

func (p *Parser) parseDataWithSchema(schema []*SchemaField) (*Value, error) {
	if err := p.expect(TokenLParen); err != nil {
		return nil, err
	}

	obj := Object()
	for i, field := range schema {
		if i > 0 {
			if p.cur.Type != TokenComma {
				return nil, fmt.Errorf("expected comma at line %d, column %d",
					p.cur.Line, p.cur.Column)
			}
			p.advance()
		}

		val, err := p.parseFieldValue(field)
		if err != nil {
			return nil, err
		}
		obj.Set(field.Name, val)
	}

	if err := p.expect(TokenRParen); err != nil {
		return nil, err
	}

	return obj, nil
}

func (p *Parser) parseFieldValue(field *SchemaField) (*Value, error) {
	if field.IsArray {
		return p.parseArrayValue(field)
	}
	if field.Children != nil {
		return p.parseDataWithSchema(field.Children)
	}
	return p.parseValue()
}

func (p *Parser) parseArrayValue(field *SchemaField) (*Value, error) {
	if err := p.expect(TokenLBracket); err != nil {
		return nil, err
	}

	arr := Array()
	first := true
	for p.cur.Type != TokenRBracket && p.cur.Type != TokenEOF {
		if !first {
			if p.cur.Type != TokenComma {
				break
			}
			p.advance()
		}
		first = false

		var val *Value
		var err error
		if field.Children != nil {
			val, err = p.parseDataWithSchema(field.Children)
		} else {
			val, err = p.parseValue()
		}
		if err != nil {
			return nil, err
		}
		arr.Push(val)
	}

	if err := p.expect(TokenRBracket); err != nil {
		return nil, err
	}

	return arr, nil
}

func (p *Parser) parseValue() (*Value, error) {
	switch p.cur.Type {
	case TokenNull:
		p.advance()
		return Null(), nil
	case TokenTrue:
		p.advance()
		return Bool(true), nil
	case TokenFalse:
		p.advance()
		return Bool(false), nil
	case TokenInteger:
		val, err := strconv.ParseInt(p.cur.Value, 10, 64)
		if err != nil {
			return nil, err
		}
		p.advance()
		return Integer(val), nil
	case TokenFloat:
		val, err := strconv.ParseFloat(p.cur.Value, 64)
		if err != nil {
			return nil, err
		}
		p.advance()
		return Float(val), nil
	case TokenString, TokenIdent:
		val := p.cur.Value
		p.advance()
		return String(val), nil
	case TokenLBracket:
		return p.parseArray()
	case TokenLParen:
		return p.parseTuple()
	default:
		return nil, fmt.Errorf("unexpected token %v at line %d, column %d",
			p.cur.Type, p.cur.Line, p.cur.Column)
	}
}

func (p *Parser) parseArray() (*Value, error) {
	if err := p.expect(TokenLBracket); err != nil {
		return nil, err
	}

	arr := Array()
	first := true
	for p.cur.Type != TokenRBracket && p.cur.Type != TokenEOF {
		if !first {
			if p.cur.Type != TokenComma {
				break
			}
			p.advance()
		}
		first = false

		val, err := p.parseValue()
		if err != nil {
			return nil, err
		}
		arr.Push(val)
	}

	if err := p.expect(TokenRBracket); err != nil {
		return nil, err
	}

	return arr, nil
}

func (p *Parser) parseTuple() (*Value, error) {
	if err := p.expect(TokenLParen); err != nil {
		return nil, err
	}

	arr := Array()
	first := true
	for p.cur.Type != TokenRParen && p.cur.Type != TokenEOF {
		if !first {
			if p.cur.Type != TokenComma {
				break
			}
			p.advance()
		}
		first = false

		val, err := p.parseValue()
		if err != nil {
			return nil, err
		}
		arr.Push(val)
	}

	if err := p.expect(TokenRParen); err != nil {
		return nil, err
	}

	return arr, nil
}

// Parse is the main entry point for parsing ASON.
func Parse(input string) (*Value, error) {
	p := NewParser(input)
	return p.Parse()
}

File: python/parser.py
"""ASON Parser."""

from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Union

from lexer import Lexer, TokenType

AsonValue = Union[None, bool, int, float, str, List[Any], Dict[str, Any]]


@dataclass
class SchemaField:
    name: str
    is_array: bool = False
    children: Optional[List["SchemaField"]] = None


class ParseError(Exception):
    """ASON parse error."""
    pass


class Parser:
    """Parses ASON input."""
    
    def __init__(self, input_str: str):
        self.lexer = Lexer(input_str)
        self.cur = self.lexer.next_token()
    
    def advance(self):
        self.cur = self.lexer.next_token()
    
    def expect(self, typ: TokenType):
        if self.cur.type != typ:
            raise ParseError(f"Expected {typ}, got {self.cur.type} at line {self.cur.line}, column {self.cur.column}")
        self.advance()
    
    def parse(self) -> AsonValue:
        """Parse the ASON input."""
        if self.cur.type == TokenType.LBRACE:
            return self.parse_schema_and_data()
        return self.parse_value()
    
    def parse_schema_and_data(self) -> AsonValue:
        schema = self.parse_schema()
        self.expect(TokenType.COLON)
        
        results = []
        while True:
            val = self.parse_data_with_schema(schema)
            results.append(val)
            
            if self.cur.type != TokenType.COMMA:
                break
            self.advance()
            
            if self.cur.type != TokenType.LPAREN:
                break
        
        return results[0] if len(results) == 1 else results
    
    def parse_schema(self) -> List[SchemaField]:
        self.expect(TokenType.LBRACE)
        
        fields = []
        while self.cur.type not in (TokenType.RBRACE, TokenType.EOF):
            field = self.parse_schema_field()
            fields.append(field)
            
            if self.cur.type == TokenType.COMMA:
                self.advance()
        
        self.expect(TokenType.RBRACE)
        return fields
    
    def parse_schema_field(self) -> SchemaField:
        if self.cur.type != TokenType.IDENT:
            raise ParseError(f"Expected identifier at line {self.cur.line}, column {self.cur.column}")
        
        field = SchemaField(name=self.cur.value)
        self.advance()
        
        if self.cur.type == TokenType.LBRACKET:
            self.advance()
            field.is_array = True
            if self.cur.type == TokenType.LBRACE:
                field.children = self.parse_schema()
            self.expect(TokenType.RBRACKET)
        elif self.cur.type == TokenType.LBRACE:
            field.children = self.parse_schema()
        
        return field
    
    def parse_data_with_schema(self, schema: List[SchemaField]) -> Dict[str, Any]:
        self.expect(TokenType.LPAREN)
        
        obj = {}
        for i, field in enumerate(schema):
            if i > 0:
                if self.cur.type != TokenType.COMMA:
                    raise ParseError(f"Expected comma at line {self.cur.line}, column {self.cur.column}")
                self.advance()
            
            val = self.parse_field_value(field)
            obj[field.name] = val
        
        self.expect(TokenType.RPAREN)
        return obj
    
    def parse_field_value(self, field: SchemaField) -> AsonValue:
        if field.is_array:
            return self.parse_array_value(field)
        if field.children:
            return self.parse_data_with_schema(field.children)
        return self.parse_value()
    
    def parse_array_value(self, field: SchemaField) -> List[Any]:
        self.expect(TokenType.LBRACKET)
        
        arr = []
        first = True
        while self.cur.type not in (TokenType.RBRACKET, TokenType.EOF):
            if not first:
                if self.cur.type != TokenType.COMMA:
                    break
                self.advance()
            first = False
            
            if field.children:
                val = self.parse_data_with_schema(field.children)
            else:
                val = self.parse_value()
            arr.append(val)
        
        self.expect(TokenType.RBRACKET)
        return arr
    
    def parse_value(self) -> AsonValue:
        t = self.cur.type
        
        if t == TokenType.NULL:
            self.advance()
            return None
        if t == TokenType.TRUE:
            self.advance()
            return True
        if t == TokenType.FALSE:
            self.advance()
            return False
        if t == TokenType.INTEGER:
            val = int(self.cur.value)
            self.advance()
            return val
        if t == TokenType.FLOAT:
            val = float(self.cur.value)
            self.advance()
            return val
        if t in (TokenType.STRING, TokenType.IDENT):
            val = self.cur.value
            self.advance()
            return val
        if t == TokenType.LBRACKET:
            return self.parse_array()
        if t == TokenType.LPAREN:
            return self.parse_tuple()

        raise ParseError(f"Unexpected token {t} at line {self.cur.line}, column {self.cur.column}")

    def parse_array(self) -> List[Any]:
        self.expect(TokenType.LBRACKET)

        arr = []
        first = True
        while self.cur.type not in (TokenType.RBRACKET, TokenType.EOF):
            if not first:
                if self.cur.type != TokenType.COMMA:
                    break
                self.advance()
            first = False
            arr.append(self.parse_value())

        self.expect(TokenType.RBRACKET)
        return arr

    def parse_tuple(self) -> List[Any]:
        self.expect(TokenType.LPAREN)

        arr = []
        first = True
        while self.cur.type not in (TokenType.RPAREN, TokenType.EOF):
            if not first:
                if self.cur.type != TokenType.COMMA:
                    break
                self.advance()
            first = False
            arr.append(self.parse_value())

        self.expect(TokenType.RPAREN)
        return arr


def parse(input_str: str) -> AsonValue:
    """Parse an ASON string and return the result."""
    p = Parser(input_str)
    return p.parse()

